### PR TITLE
Optimize view traversal

### PR DESCRIPTION
**Pull request by noamnelke**

### CODE REVIEW

File: mesh/meshdb.go
Comment (almogdepaz) [inline mesh/meshdb.go]: already said that i don't think this is a good idea

File: mesh/meshdb.go
Comment (almogdepaz) [inline mesh/meshdb.go:145]: dont put blocks that are to deep in the queue in the first place
check this in the view edges loop

File: mesh/meshdb.go
Comment (almogdepaz) [inline mesh/meshdb.go]: already said that i don't think this is a good idea,
if you do this please change all uses of list with your implementation

File: mesh/meshdb.go
Comment (noamnelke) [reply mesh/meshdb.go:145]: This is an intentional change and tradeoff. The alternative is to get the block from the cache twice (once to check the layer and another to perform processing, incl. getting the children).

The cost of checking this at this point is that we need to enter the loop more times. Specifically, we're expected to enter the loop about 200 more times (an additional layer) but save `200 * <layers traversed>` calls to `getBlock()`. Empirically speaking, it results in reduced runtime in the benchmark (10-15%).

File: mesh/meshdb.go
Comment (noamnelke) [reply mesh/meshdb.go]: reverted this change d73734e4

File: mesh/meshdb.go
Comment (almogdepaz) [reply mesh/meshdb.go:145]: the blocks in the second "GetBlock" are the blocks from the view so what you say about calling it twice on the same block  is incorrect 
i think validating the blocks before inserting them to the queue is cleaner

Comment (noamnelke) [reply mesh/meshdb.go:145]: "the blocks from the view" are what gets put into the queue and then we call `GetBlock()` on them when it's their turn to be taken out of the queue.

I agree that it would be cleaner, but I don't think it's worth a 10-15% performance hit.

Comment (almogdepaz) [reply mesh/meshdb.go]: a better way to fix it is to out pointers to blocks in the queue instead of id's

Comment (almogdepaz) [reply mesh/meshdb.go:145]: in your solution the queue will hold blocks that are not needed at all 
maybe a better way to fix it is to put pointers to blocks in the queue instead of id's
keep the same flow -> avoid the double get

Comment (noamnelke) [reply mesh/meshdb.go:145]: This was my initial thought as well and if you look at the commits in this PR you'll find this commit that does that: https://github.com/spacemeshos/go-spacemesh/pull/1212/commits/cdb68faa74f4f6593f67f8194ca0f81234cfe001

I later reverted this change and changed the flow because I came to the realization that the memory footprint of keeping all these blocks would be huge:

200 blocks per layer
2 layers in the queue (after handling the first block you should have all of the next layer queued, while still having most of the first layer there)
Let's take an optimistic approach and say a mini block is 10kb.

400 blocks * 10kb = 4mb

This is a very conservative calculation and it's still too big to run this in multiple threads concurrently. WDYT?

**Review by zalmen — APPROVED** [review]

File: mesh/meshdb.go
Comment (almogdepaz) [reply mesh/meshdb.go:145]: these blocks are already in the cache you are just holding a pointer i dont agree with the reasons or the calculations you gave

Comment (almogdepaz) [reply mesh/meshdb.go:145]: besides you know that the all the block ids in the view are from layers older then the block you are working on you don't need to pull them from the cache to figure out if you should insert the id in the queue.
correct me if im wrong but in this way you get only one call to getBlock
still use block ids in the queue  and the condition can stay before the insertion.

Comment (noamnelke) [reply mesh/meshdb.go:145]: Great point! I've added this check now in 6f857771. However, we still need the check before the handler because blocks could reference blocks more than one layer back and then blocks could "slip through".

If you still don't like the way it's implemented then let's discuss this in person tomorrow.

Comment (almogdepaz) [reply mesh/meshdb.go:145]: nope you did just what i meant it's great

Comment (noamnelke) [reply mesh/meshdb.go:145]: Perfect! Yuval asked me to get an approval from you before merging -- any other reservations?

Comment (almogdepaz) [reply mesh/meshdb.go:145]: thanks, already approved it

**Comment by noamnelke** [comment]
✅ Automation tests passed

(Had to increase `gossip_propagation_sleep` in `test_many_gossip_messages` from 10 to 50 seconds 😕 it failed with 35)

**Pull request merged**
